### PR TITLE
Ask to join meeting

### PR DESCRIPTION
**Pull request by RemcoSimonides**
- [x] User can ask to join
ask to join

- [x] Organizer can accept or deny requests - multiple people can request at the same time
multiple people notification

- [x] User can request again after being denied
- [x] User is automatically navigated to session when accepted

### CODE REVIEW

File: apps/festival/festival/src/app/marketplace/event/session/session.component.ts [hunk 1]
```diff
@@ -42,27 +50,42 @@ export class SessionComponent implements OnInit, OnDestroy {
 
       if (event.type === 'screening') {
         if (!!(event.meta as Screening).titleId) {
-          const movie = await this.movieService.getValue(event.meta.titleId as string)
+          const movie = await this.movieService.getValue(event.meta.titleId as string);
           this.screeningFileRef = movie.promotional.videos?.screener?.ref ?? '';
         }
       } else if (event.type === 'meeting') {
         if (event.isOwner) {
```
Comment (pldespaigne) [inline apps/festival/festival/src/app/marketplace/event/session/session.component.ts:57]: `} else if (event.type === 'meeting' && event.isOwner) {` as you did line 82

File: apps/festival/festival/src/app/marketplace/event/session/session.component.ts [hunk 2]
```diff
@@ -42,27 +50,42 @@ export class SessionComponent implements OnInit, OnDestroy {
...
+  attendeeLeaves() {
     if (this.event.isOwner && this.event.type === 'meeting') {
-      (this.event.meta as Meeting).ownerIsPresent = false;
-      this.service.update(this.event.id, { meta: this.event.meta });
+      const meta: Meeting = { ...this.event.meta, attendees: {} };
+      this.service.update(this.event.id, { meta });
+    } else if (this.event.type === 'meeting') {
+      const meta = { ...this.event.meta, attendees: {...this.event.meta.attendees} };
+      if (!!meta.attendees[this.authQuery.userId]) delete meta.attendees[this.authQuery.userId];
+      this.service.update(this.event.id, { meta });
```
Comment (pldespaigne) [inline apps/festival/festival/src/app/marketplace/event/session/session.component.ts:88]:
```TypeScript
if (this.event.type === 'meeting') {
  const meta = { ...this.event.meta, attendees: {...this.event.meta.attendees} };
  if (this.event.isOwner) {
    meta.attendees = {};
  } else if (!!meta.attendees[this.authQuery.userId]){
    delete meta.attendees[this.authQuery.userId];
  }
  this.service.update(this.event.id, { meta });
}
```

File: libs/event/src/lib/+state/event.firestore.ts [hunk 3]
```diff
@@ -7,11 +7,13 @@ type Timestamp = firestore.Timestamp;
...
 export type EventMeta = Meeting | Screening | {};
 
+export type AttendeeStatus = 'owner' | 'requesting' | 'accepted' | 'denied';
+
 export interface Meeting {
   organizerId: string;
   description: string;
   files: string[];
-  ownerIsPresent: boolean;
+  attendees: Record<string, AttendeeStatus>;
```
Comment (pldespaigne) [inline libs/event/src/lib/+state/event.firestore.ts:16]: please move attendees above files so everything is nicely grouped ^^

File: libs/event/src/lib/+state/event.model.ts [hunk 4]
```diff
@@ -58,8 +58,7 @@ export function createMeeting(meeting: Partial<Meeting>): Meeting {
     organizerId: '',
     description: '',
     files: [],
-    ownerIsPresent: false,
-
+    attendees: {},
```
Comment (pldespaigne) [inline libs/event/src/lib/+state/event.model.ts:61]: same

File: libs/event/src/lib/components/doorbell/doorbell.component.html [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<ng-container *ngFor="let requestor of data.requests">
+  <section fxLayout fxLayoutAlign="space-between center">
+    <div>{{ requestor | displayUser | async }} wants to join</div>
```
Comment (pldespaigne) [inline libs/event/src/lib/components/doorbell/doorbell.component.html:3]: this should be a `<p>`

**Review by pldespaigne — CHANGES_REQUESTED** [review]

File: apps/festival/festival/src/app/marketplace/event/lobby/lobby.component.ts [hunk 6]
```diff
@@ -34,13 +44,33 @@ export class LobbyComponent implements OnInit, OnDestroy {
...
     const name = displayName(this.authQuery.user);
     this.twilioService.initLocal(name);
+
+    this.sub = this.event$.subscribe(event => {
+      if (event.type === 'meeting') {
+        const attendees = (event.meta as Meeting).attendees;
+        const values = Object.keys(attendees).map(key => attendees[key]);
+        const uid = this.authQuery.userId;
+        this.ownerIsPresent = values.some(value => value === 'owner');
+        this.attendeeStatus = event.isOwner ? 'owner' : attendees[uid];
```
[the remark this answers is not in the document]
Comment (RemcoSimonides) [reply apps/festival/festival/src/app/marketplace/event/lobby/lobby.component.ts:54]: Actually it is not. If the owner goes to the lobby, he hasn't set `attendees[uid] = 'owner'` yet; that happens when the owner goes to the session page.
However, the button to go to the session page is on a `[ngSwitch]="attendeeStatus"`. If the case is 'owner', the button is enabled. I could move the button for owner outside the ngSwitch

```html
<ng-container [ngSwitch]="attendeeStatus">
        <ng-container *ngSwitchCase="'owner' || 'accepted'">
          <a mat-flat-button routerLink="../session" color="accent" class="request">Access {{ event.type | titlecase }} Room</a>
        </ng-container>
        <ng-container *ngSwitchCase="'requesting'">
          <a mat-flat-button disabled color="accent" class="request">Waiting for request to be accepted...</a>
        </ng-container>
        <ng-container *ngSwitchCase="'denied'">
          <div fxLayout="column" fxLayoutAlign="end center" class="request">
            Your request to join has been denied.
            <a mat-flat-button (click)="requestAccess(event)" color="accent">Ask to join again</a>
          </div>
        </ng-container>
        <ng-container *ngSwitchDefault>
          <ng-container *ngIf="ownerIsPresent else ownerNotPresent">
            <a mat-flat-button (click)="requestAccess(event)" color="accent" class="request">Ask to join {{ event.type | titlecase }} Room</a>
          </ng-container>
          <ng-template #ownerNotPresent>
            <a mat-flat-button disabled color="accent" class="request">Waiting for owner...</a>
          </ng-template>
        </ng-container>
      </ng-container>
```

File: apps/festival/festival/src/app/marketplace/event/lobby/lobby.component.ts [hunk 7]
```diff
@@ -34,13 +44,33 @@ export class LobbyComponent implements OnInit, OnDestroy {
...
   ngOnDestroy() {
     this.twilioService.cleanLocal();
+    this.sub.unsubscribe();
   }
 
   toggleLocalTrack(kind: TrackKind) {
     this.twilioService.toggleTrack(kind);
   }
+
+  requestAccess(event: Event) {
```
[the remark this answers is not in the document]
Comment (RemcoSimonides) [reply apps/festival/festival/src/app/marketplace/event/lobby/lobby.component.ts:71]: So in this case I get it from `eventQuery.getActive()` instead?

File: apps/festival/festival/src/app/marketplace/event/lobby/lobby.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pldespaigne) [reply apps/festival/festival/src/app/marketplace/event/lobby/lobby.component.ts]: ok my bad, that was a bit complex to get at first look, a comment could have helped ^^

**Review by pldespaigne — APPROVED** [review]

**Pull request merged**